File: experiment/backends/ios/sqlitevectorai/SqliteVectorAiVector.swift
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Foundation

/// Benchmark wrapper around the SQLiteAI `sqlite-vector` extension on iOS.
/// Mirrors experiment/backends/android/core/SqliteVectorAiVector.kt so
/// the cross-platform vector bench runs the exact same per-query shape on
/// both platforms (ingest → vector_quantize → vector_quantize_scan).
///
/// The extension binary is Elastic License 2.0 — benchmark-only, never
/// distributed inside the dazzle SDK. This file + the sqlitevectorai/
/// bridge are only compiled into the DazzleBackends experiment target.
final class SqliteVectorAiVector {

    private let dim: Int
    private let dbPath: String
    private var handle: OpaquePointer?
    // String id ↔ integer rowid mapping. `add(id:vector:)` takes a string
    // so the bench's recall comparison sees the same id space the other
    // backends return. sqlite-vector expects an INTEGER PRIMARY KEY, so we
    // keep a parallel array and translate back on search.
    private var idToRowid: [String: Int64] = [:]
    private var rowidToId: [String] = []

    init(dim: Int, dbName: String = "vecbench_sqlvectorai") {
        self.dim = dim
        let caches = FileManager.default.urls(for: .cachesDirectory,
                                              in: .userDomainMask)[0]
        self.dbPath = caches.appendingPathComponent("\(dbName).db").path
    }

    /// Path to the embedded vector.framework binary. The bundled
    /// (vendored) sqlite3 inside svai_ios.c calls `sqlite3_load_extension`
    /// with this absolute path; iOS resolves it via dlopen at runtime.
    /// Returns nil if the framework was not embedded — that is a config
    /// bug, not a runtime fallback path; callers should treat it as a
    /// hard skip of the SqliteVectorAi bench leg.
    private static func extensionBinaryPath() -> String? {
        // PrivateFrameworks dir is where Xcode embeds vector.xcframework
        // for `embed: true, codeSign: true`. On both device and simulator
        // bundles the Mach-O lives at <Frameworks>/vector.framework/vector.
        guard let fwBase = Bundle.main.privateFrameworksPath else {
            return nil
        }
        let candidate = "\(fwBase)/vector.framework/vector"
        guard FileManager.default.fileExists(atPath: candidate) else {
            return nil
        }
        return candidate
    }

    /// Returns true on success, false if the SQLiteAI extension couldn't
    /// be loaded. The most common failures here are (a) vector.framework
    /// not embedded into the bundle, or (b) the bundled sqlite3
    /// translation unit failed to compile in `SQLITE_ENABLE_LOAD_EXTENSION`
    /// mode — both are config bugs caught at first run, not transient.
    @discardableResult
    func create() -> Bool {
        try? FileManager.default.removeItem(atPath: dbPath)
        try? FileManager.default.removeItem(atPath: dbPath + "-wal")
        try? FileManager.default.removeItem(atPath: dbPath + "-shm")
        idToRowid.removeAll()
        rowidToId.removeAll()

        guard let extPath = Self.extensionBinaryPath() else {
            print("[svai] vector.framework binary not found in app bundle")
            return false
        }

        handle = dbPath.withCString { cdb -> OpaquePointer? in
            extPath.withCString { cext -> OpaquePointer? in
                svai_open(cdb, cext, Int32(dim))
            }
        }
        return handle != nil
    }

    func close() {
        if let h = handle { svai_close(h); handle = nil }
    }

    func add(id: String, vector: [Float]) {
        precondition(vector.count == dim, "vector size \(vector.count) != dim \(dim)")
        guard let h = handle else { return }
        // SQLite rowid is 1-indexed. Using size+1 keeps inserts monotonic
        // which helps the vector_quantize snapshot be stable.
        let rowid = Int64(rowidToId.count) + 1
        vector.withUnsafeBufferPointer { vp in
            svai_add(h, rowid, vp.baseAddress, Int32(dim))
        }
        idToRowid[id] = rowid
        rowidToId.append(id)
    }

    /// Bulk ingest wrapped in a single SQLite transaction, then builds the
    /// quantized snapshot via `vector_quantize` — mandatory before any
    /// search per SQLiteAI API.md.
    func addAll(ids: [String], vectors: [[Float]]) {
        precondition(ids.count == vectors.count,
                     "ids and vectors must be parallel")
        guard let h = handle else { return }
        svai_begin_tx(h)
        for i in 0..<ids.count {
            add(id: ids[i], vector: vectors[i])
        }
        svai_commit_tx(h)
    }

    /// Build the quantized snapshot. MUST be called after the last `add` /
    /// `addAll` and before the first `search`; without it
    /// `vector_quantize_scan` returns zero rows.
    @discardableResult
    func finalizeIndex(maxMemoryMb: Int = 50, preload: Bool = true) -> Bool {
        guard let h = handle else { return false }
        let rc = svai_finalize_index_ex(h, Int32(maxMemoryMb), preload ? 1 : 0)
        if rc != 0 {
            print("[svai] vector_quantize failed rc=\(rc)")
            return false
        }
        return true
    }

    /// k-NN scan. Returns (id, distance) pairs sorted ascending.
    func search(query: [Float], k: Int) -> [(id: String, distance: Float)] {
        precondition(query.count == dim, "query size \(query.count) != dim \(dim)")
        guard let h = handle else { return [] }

        var outRowids = [Int64](repeating: 0, count: k)
        var outDists  = [Float](repeating: 0, count: k)

        let n: Int32 = query.withUnsafeBufferPointer { qp in
            outRowids.withUnsafeMutableBufferPointer { idsP in
                outDists.withUnsafeMutableBufferPointer { distsP in
                    svai_knn(h, qp.baseAddress, Int32(dim), Int32(k),
                             idsP.baseAddress, distsP.baseAddress, Int32(k))
                }
            }
        }

        var out: [(id: String, distance: Float)] = []
        out.reserveCapacity(Int(n))
        for i in 0..<Int(n) {
            let rowid = outRowids[i]
            let idx = Int(rowid - 1)
            if idx >= 0 && idx < rowidToId.count {
                out.append((id: rowidToId[idx], distance: outDists[i]))
            }
        }
        return out
    }

    func count() -> Int64 {
        guard let h = handle else { return -1 }
        return svai_count(h)
    }

    func dbFileSizeBytes() -> Int64 {
        dbPath.withCString { svai_db_file_size($0) }
    }
}

File: experiment/backends/android/core/SqliteVectorAiVector.kt
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
// SPDX-License-Identifier: Apache-2.0

package dev.dazzle.experiment

import android.content.Context
import java.io.File
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * SQLiteAI `sqlite-vector` backend for the vector-search benchmark.
 *
 * Statically-linked SQLite 3.x + `libvector.so` (ai.sqlite:vector:0.9.80,
 * extracted from the public Maven AAR at build time and placed in
 * jniLibs/arm64-v8a/). The extension is Elastic License 2.0 — bench-only.
 *
 * Table layout: `emb(id INTEGER PRIMARY KEY, vector BLOB)`. String document
 * ids come in from the test harness; we map them 1:1 to the int rowid by
 * keeping a parallel ArrayList<String>, so the harness's recall comparison
 * still sees the same id space the other backends return.
 *
 * Search is brute-force SIMD (sqlite-vector ships no ANN index by default);
 * recall is always 1.0 by construction, so what the bench really measures
 * here is per-query scan wall-time vs dazzle-vector's HNSW.
 */
class SqliteVectorAiVector(
    private val context: Context,
    private val dim: Int,
    private val dbName: String = "vecbench_sqlvectorai",
) {

    private var handle: Long = 0
    private val idToRowid = HashMap<String, Long>()
    private val rowidToId = ArrayList<String>()

    fun create() {
        val file = File(context.filesDir, "$dbName.db")
        if (file.exists()) file.delete()
        File(context.filesDir, "$dbName.db-wal").delete()
        File(context.filesDir, "$dbName.db-shm").delete()
        idToRowid.clear()
        rowidToId.clear()
        handle = nOpen(file.absolutePath, dim)
        check(handle != 0L) { "sqlite-vector open failed — see logcat SqliteVectorAIJNI" }
    }

    fun close() {
        if (handle != 0L) { nClose(handle); handle = 0 }
    }

    private fun vecToDirect(v: FloatArray): ByteBuffer {
        val buf = ByteBuffer.allocateDirect(v.size * 4).order(ByteOrder.LITTLE_ENDIAN)
        buf.asFloatBuffer().put(v)
        buf.position(0)
        return buf
    }

    fun add(id: String, vector: FloatArray) {
        require(vector.size == dim)
        val rowid = rowidToId.size.toLong() + 1L  // SQLite rowid is 1-indexed
        nAdd(handle, rowid, vecToDirect(vector))
        idToRowid[id] = rowid
        rowidToId += id
    }

    /** Bulk ingest wrapped in a single SQLite transaction. */
    fun addAll(ids: Array<String>, vectors: Array<FloatArray>) {
        require(ids.size == vectors.size)
        nBeginTx(handle)
        try {
            for (i in ids.indices) {
                val rowid = rowidToId.size.toLong() + 1L
                nAdd(handle, rowid, vecToDirect(vectors[i]))
                idToRowid[ids[i]] = rowid
                rowidToId += ids[i]
            }
            nCommitTx(handle)
        } catch (t: Throwable) {
            nCommitTx(handle)
            throw t
        }
    }

    /** k-NN search. Returns (id, distance) sorted by ascending distance. */
    fun search(query: FloatArray, k: Int): List<Pair<String, Float>> {
        require(query.size == dim)
        @Suppress("UNCHECKED_CAST")
        val pair = nKnn(handle, vecToDirect(query), k) ?: return emptyList()
        val ids   = pair[0] as LongArray
        val dists = pair[1] as FloatArray
        val out = ArrayList<Pair<String, Float>>(ids.size)
        for (i in ids.indices) {
            val rowid = ids[i]
            val idx = (rowid - 1L).toInt()
            if (idx in rowidToId.indices) out += rowidToId[idx] to dists[i]
        }
        return out
    }

    /**
     * Build the quantized snapshot over `emb.vector`. Must be called after
     * `addAll` / the last `add`, before the first `search`. Without it,
     * `vector_quantize_scan` returns zero rows (observed: recall=0).
     */
    fun finalizeIndex(maxMemoryMb: Int = 50, preload: Boolean = true) {
        val rc = nFinalizeIndex(handle, maxMemoryMb, preload)
        check(rc == 0) { "vector_quantize failed rc=$rc — see logcat SqliteVectorAIJNI" }
    }

    fun count(): Long = nCount(handle)

    fun dbFileSizeBytes(): Long {
        val f = File(context.filesDir, "$dbName.db")
        return if (f.exists()) f.length() else -1L
    }

    companion object {
        init {
            // libvector.so must be loaded first so its symbols are resolvable
            // when sqlite3_load_extension(dlopen) fires inside nOpen.
            try { System.loadLibrary("vector") } catch (_: UnsatisfiedLinkError) {}
            System.loadLibrary("sqlitevectorai-jni")
        }

        @JvmStatic private external fun nOpen(path: String, dim: Int): Long
        @JvmStatic private external fun nClose(handle: Long)
        @JvmStatic private external fun nBeginTx(handle: Long)
        @JvmStatic private external fun nCommitTx(handle: Long)
        @JvmStatic private external fun nAdd(handle: Long, id: Long, vec: ByteBuffer)
        @JvmStatic private external fun nKnn(handle: Long, query: ByteBuffer, k: Int): Array<Any>?
        @JvmStatic private external fun nFinalizeIndex(
            handle: Long,
            maxMemoryMb: Int,
            preload: Boolean,
        ): Int
        @JvmStatic private external fun nCount(handle: Long): Long
    }
}
